tests/unit/test_parallel_write.py

```diff
@@ -0,0 +1,57 @@
+"""Module for testing the parallel write feature for the ZarrIO."""
+from pathlib import Path
+from typing import Tuple, Dict
+
+import numpy as np
+from hdmf_zarr import ZarrIO
+from hdmf.common import DynamicTable, VectorData
+from hdmf.data_utils import GenericDataChunkIterator
+
+
+class TestDataChunkIterator(GenericDataChunkIterator):
+    """Generic data chunk iterator used for specific testing purposes."""
+
+    def __init__(self, data, **base_kwargs):
+        self.data = data
+
+        self._base_kwargs = base_kwargs
+        super().__init__(**base_kwargs)
+
+    def _get_dtype(self) -> np.dtype:
+        return self.data.dtype
+
+    def _get_maxshape(self) -> tuple:
+        return self.data.shape
+
+    def _get_data(self, selection: Tuple[slice]) -> np.ndarray:
+        return self.data[selection]
+
+    def __reduce__(self):
+        instance_constructor = self._from_dict
+        initialization_args = (self._to_dict(),)
+        return (instance_constructor, initialization_args)
+
+    def _to_dict(self) -> Dict:
+        dictionary = dict()
+        # Note this is not a recommended way to pickle contents
+        # ~~ Used for testing purposes only ~~
+        dictionary["data"] = self.data
+        dictionary["base_kwargs"] = self._base_kwargs
+
+        return dictionary
+
+    @staticmethod
+    def _from_dict(dictionary: dict) -> GenericDataChunkIterator:  # TODO: need to investigate the need of base path
+        source_type = dictionary["data"]
+
+        iterator = TestDataChunkIterator(data=data, **dictionary["base_kwargs"])
+        return iterator
+
+def test_parallel_write(tmpdir):
+    number_of_jobs = 2
+    column = VectorData(name="TestColumn", description="", data=TestDataChunkIterator(data=np.array([1., 2., 3.])))
+    dynamic_table = DynamicTable(name="TestTable", description="", columns=[column])
+
+    zarr_top_level_path = str(tmpdir / f"example_parallel_zarr_{number_of_jobs}.zarr")
+    with ZarrIO(path=zarr_top_level_path, mode="w") as io:
+        io.write(dynamic_table, number_of_jobs=number_of_jobs)
```

Comment:
First attempt at just writing a DynamicTable object alone gives the following error stack

```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[9], line 56
     54 zarr_top_level_path = str(tmpdir / f"example_parallel_zarr_{number_of_jobs}.zarr")
     55 with ZarrIO(path=zarr_top_level_path, mode="w") as io:
---> 56     io.write(dynamic_table, number_of_jobs=number_of_jobs)

File ~/GitHub/hdmf/src/hdmf/utils.py:644, in docval.<locals>.dec.<locals>.func_call(*args, **kwargs)
    642 def func_call(*args, **kwargs):
    643     pargs = _check_args(args, kwargs)
--> 644     return func(args[0], **pargs)

File ~/GitHub/hdmf-zarr/src/hdmf_zarr/backend.py:231, in ZarrIO.write(self, **kwargs)
    229 """Overwrite the write method to add support for caching the specification"""
    230 cache_spec = popargs('cache_spec', kwargs)
--> 231 super(ZarrIO, self).write(**kwargs)
    232 if cache_spec:
    233     self.__cache_spec()

File ~/GitHub/hdmf/src/hdmf/utils.py:644, in docval.<locals>.dec.<locals>.func_call(*args, **kwargs)
    642 def func_call(*args, **kwargs):
    643     pargs = _check_args(args, kwargs)
--> 644     return func(args[0], **pargs)

File ~/GitHub/hdmf/src/hdmf/backends/io.py:81, in HDMFIO.write(self, **kwargs)
     79 """Write a container to the IO source."""
     80 container = popargs('container', kwargs)
---> 81 f_builder = self.__manager.build(container, source=self.__source, root=True)
     82 self.write_builder(f_builder, **kwargs)
     84 if self.external_resources_path is not None:

File ~/GitHub/hdmf/src/hdmf/utils.py:644, in docval.<locals>.dec.<locals>.func_call(*args, **kwargs)
    642 def func_call(*args, **kwargs):
    643     pargs = _check_args(args, kwargs)
--> 644     return func(args[0], **pargs)

File ~/GitHub/hdmf/src/hdmf/build/manager.py:171, in BuildManager.build(self, **kwargs)
    167             raise ValueError("Cannot change container_source once set: '%s' %s.%s"
    168                              % (container.name, container.__class__.__module__,
    169                                 container.__class__.__name__))
    170 # NOTE: if exporting, then existing cached builder will be ignored and overridden with new build result
--> 171 result = self.__type_map.build(container, self, source=source, spec_ext=spec_ext, export=export)
    172 self.prebuilt(container, result)
    173 self.__active_prebuilt(result)

File ~/GitHub/hdmf/src/hdmf/utils.py:644, in docval.<locals>.dec.<locals>.func_call(*args, **kwargs)
    642 def func_call(*args, **kwargs):
    643     pargs = _check_args(args, kwargs)
--> 644     return func(args[0], **pargs)

File ~/GitHub/hdmf/src/hdmf/build/manager.py:763, in TypeMap.build(self, **kwargs)
    760 source, spec_ext, export = getargs('source', 'spec_ext', 'export', kwargs)
    762 # get the ObjectMapper to map between Spec objects and AbstractContainer attributes
--> 763 obj_mapper = self.get_map(container)
    764 if obj_mapper is None:
    765     raise ValueError('No ObjectMapper found for container of type %s' % str(container.__class__.__name__))

File ~/GitHub/hdmf/src/hdmf/utils.py:644, in docval.<locals>.dec.<locals>.func_call(*args, **kwargs)
    642 def func_call(*args, **kwargs):
    643     pargs = _check_args(args, kwargs)
--> 644     return func(args[0], **pargs)

File ~/GitHub/hdmf/src/hdmf/build/manager.py:705, in TypeMap.get_map(self, **kwargs)
    703     namespace, data_type = self.get_container_cls_dt(container_cls)
    704     if namespace is None:
--> 705         raise ValueError("class %s is not mapped to a data_type" % container_cls)
    706 else:
    707     data_type = self.get_builder_dt(obj)

ValueError: class <class 'hdmf.common.table.DynamicTable'> is not mapped to a data_type
```